src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -898,13 +898,14 @@ void ImageReplayer<I>::shut_down(int r) {
   // chain the shut down sequence (reverse order)
   Context *ctx = new LambdaContext(
     [this, r](int _r) {
-      update_mirror_image_status(true, STATE_STOPPED);
       handle_shut_down(r);
     });
 
   // destruct the state builder
   if (m_state_builder != nullptr) {
     ctx = new LambdaContext([this, ctx](int r) {
+      set_state_description(0, "local image is primary");
+      update_mirror_image_status(true, STATE_STOPPED);
```

Comment:
While reporting a `STOPPING` replayer as `STOPPED` is not _that_ big a deal, this is really just hiding a symptom.  I would rather we attack the actual issue which boils down to image replayer getting irreparably stuck if the primary cluster ceases to exist due to a disaster.  It's not limited to `aio_unwatch()` -- any librados call is susceptible to this.

I would look in the direction of setting `rados_osd_op_timeout` for the remote RADOS client to a configurable value (e.g. `rbd_mirror_remote_osd_op_timeout` option).  This should help not just here but also in all kinds of force promote hang scenarios which @chrisphoffman has been looking at.

Reply:
Yes this is somthing that I was asking on the other day, how to avoid the stuck phase, but other topic carried away soon. Thanks @idryomov 

BTW - I hope we can use:
```
# ceph --cluster site-a config set global rados_osd_op_timeout 10
# ceph --cluster site-b config set global rados_osd_op_timeout 10
```
for now.

With this the LambdaContext should be invoked,

```
Context *ctx = new LambdaContext(
┆ [this, r](int _r) {            
┆ ┆ update_mirror_image_status(true, STATE_STOPPED);
┆ ┆ handle_shut_down(r);         
┆ });                            
```

But we should still call `set_state_description(0, "local image is primary"); ` else the status will be `up+error` and with description `local image linked to unknown peer`.

If calling `set_state_description(0, "local image is primary"); `  in shut_down() doesn't sound good, do you have a suggestion on it?

Reply:
I don't think any additional calls to `set_state_description()` are needed -- it really seems like the "local image linked to unknown peer" description is generated wrongfully.  I suspect the image replayer produces it upon encountering an orphan snapshot and we just need to tweak the corresponding logic in `Replayer::scan_local_mirror_snapshots()`.

Reply:
Yeah, I'm actually on it :-)

Reply:
> BTW - I hope we can use:
> 
> ```
> # ceph --cluster site-a config set global rados_osd_op_timeout 10
> ```

Only for development/testing purposes.

Reply:
Sure, I will let you know once I have something. I was tracing back the stuff and saw 

```
if (m_local_snap_id_start > 0 || m_local_snap_id_end != CEPH_NOSNAP) {
┆ if (m_local_mirror_snap_ns.is_non_primary() &&                      
┆ ┆ ┆ m_local_mirror_snap_ns.primary_mirror_uuid !=                   
┆ ┆ ┆ ┆ m_state_builder->remote_mirror_uuid) {                        
┆ ┆ // TODO support multiple peers                                    
┆ ┆ derr << "local image linked to unknown peer: "                    
┆ ┆ ┆ ┆ ┆<< m_local_mirror_snap_ns.primary_mirror_uuid << dendl;      
┆ ┆ handle_replay_complete(locker, -EEXIST,                           
┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆"local image linked to unknown peer");     
┆ ┆ return;                                                           
┆ } else if (m_local_mirror_snap_ns.state ==                          
┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {             
┆ ┆ dout(5) << "local image promoted" << dendl;                       
┆ ┆ handle_replay_complete(locker, 0, "force promoted");              
┆ ┆ return;                                                           
┆ }                                                                   

```

Thanks again!

Reply:
> Only for development/testing purposes.

Sorry could you elaborate on the topic please,
Are you saying we should have an option at rbd-mirroing level `rbd_mirror_remote_osd_op_timeout` to tweak `rados_osd_op_timeout` for the mirroring daemons?

Reply:
> Are you saying we should have an option at rbd-mirroing level `rbd_mirror_remote_osd_op_timeout` to tweak `rados_osd_op_timeout` for the mirroring daemons?

Yes, but not for the entire rbd-mirror daemon -- only for its "remote" RADOS client(s).  I even provided a suggestion for the option name ;)

> I would look in the direction of setting `rados_osd_op_timeout` for the remote RADOS client to a configurable value (e.g. `rbd_mirror_remote_osd_op_timeout` option).